src/libmongoc/src/mongoc/mongoc-trace-private.h

```diff
@@ -28,53 +28,82 @@
+      mlib_diagnostic_pop ();                                   \
+   } else                                                       \
+      ((void) 0)
+
 #define ENTRY                                                                                                   \
-   do {                                                                                                         \
+   if (1) {                                                                                                     \
+      mlib_diagnostic_push ();                                                                                  \
+      mlib_disable_constant_conditional_expression_warnings ();                                                 \
       if (MONGOC_TRACE_ENABLED && gLogTrace) {                                                                  \
          mongoc_log (MONGOC_LOG_LEVEL_TRACE, MONGOC_LOG_DOMAIN, "ENTRY: %s():%d", BSON_FUNC, (int) (__LINE__)); \
       }                                                                                                         \
-   } while (0)
+      mlib_diagnostic_pop ();                                                                                   \
+   } else                                                                                                       \
+      ((void) 0)
+
 #define EXIT                                                                                                    \
-   do {                                                                                                         \
+   if (1) {                                                                                                     \
+      mlib_diagnostic_push ();                                                                                  \
+      mlib_disable_constant_conditional_expression_warnings ();                                                 \
       if (MONGOC_TRACE_ENABLED && gLogTrace) {                                                                  \
          mongoc_log (MONGOC_LOG_LEVEL_TRACE, MONGOC_LOG_DOMAIN, " EXIT: %s():%d", BSON_FUNC, (int) (__LINE__)); \
       }                                                                                                         \
       return;                                                                                                   \
-   } while (0)
+      mlib_diagnostic_pop ();                                                                                   \
+   } else                                                                                                       \
+      ((void) 0)
+
 #define RETURN(ret)                                                                                             \
-   do {                                                                                                         \
+   if (1) {                                                                                                     \
```

Comment:
Heads up: this may need to be tweaked. Seems to be triggering a warning on the [mock-server-test](https://spruce.mongodb.com/task/mongo_c_driver_mock_server_test_mock_server_test_patch_66f23a9d463e842510fdb8f335ac0d8da1907a31_67f6ce212755bd0007484205_25_04_09_19_44_36/logs?execution=0) task:

```
mongoc-aggregate.c:318:1: error: control reaches end of non-void function [-Werror=return-type]
```

Reply:
Yep, investigating. It appears to be a false-positive where it is not observing the `return` within the branch despite the `if (1)` constant conditional.

Reply:
Looks like the if-else pattern triggers false-positives in multiple contexts, not just in `mongoc-aggregate.c` with GCC 11. Reverted trace macros containing `return` statements to their former do-while form to avoid triggering this issue.